htlcswitch/payment_result.go

```diff
@@ -0,0 +1,47 @@
+package htlcswitch
+
+import (
+	"errors"
+
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// ErrPaymentIDNotFound is an error returned if the given paymentID is
+	// not found.
+	ErrPaymentIDNotFound = errors.New("paymentID not found")
+
+	// ErrPaymentIDAlreadyExists is returned if we try to write a pending
+	// payment whose paymentID already exists.
+	ErrPaymentIDAlreadyExists = errors.New("paymentID already exists")
+)
+
+// PaymentResult wraps a decoded result received from the network after a
+// payment attempt was made. This is what is eventually handed to the router
+// for processing.
+type PaymentResult struct {
+	// Preimage is set by the switch in case a sent HTLC was settled.
+	Preimage [32]byte
+
+	// Error is non-nil in case a HTLC send failed, and the HTLC is now
+	// irrevocably cancelled.
+	Error *ForwardingError
+}
+
+// networkResult is the raw result received from the network after a payment
+// attempt has been made. Since the switch doesn't always have the necessary
+// data to decode the raw message, we store it together with some meta data,
+// and decode it when the router query for the final result.
+type networkResult struct {
+	// msg is the received result. This should be of type UpdateFulfillHTLC
+	// or UpdateFailHTLC.
+	msg lnwire.Message
+
+	// unencrypted indicates whether the failure encoded in the message is
+	// unencrypted, and hence doesn't need to be decrypted.
+	unencrypted bool
+
+	// isResolution indicates whether this is a resolution message, in
+	// which the failure reason might not be included.
```

Comment:
IIRC we return "permanent chan failure" if we need to sweep it on chain, then cancel back along the incoming link.

Reply:
I see at least one place where the failure reason is non-nil: https://github.com/lightningnetwork/lnd/blob/a8fa4094ba61e26bd9855e9743883e3a234d383a/htlcswitch/switch.go#L1184

cc @cfromknecht

Reply:
In that line, it isn't a locally sourced payment. So that reason can not end up as a payment result in the switch with `isResolution` true?

Reply:
I think it would be safe to remove the `isResolution` bool, and just leave the `htlc.Reason==nil` check (or safer `len(htlc.Reason)==0` to handle serialization`)? If there are other cases where the reason can be nil, that is not resolution, decryption would already be failing...

Reply:
Decided to keep this as is for now, and decide in the persistence PR.

Reply:
:+1: